Add causal LM for ASR2

akreal opened the pull request:
This PR also includes few things which are not directly related to the causal LM, but are needed for the recipe:

- add an option to disable bias in the CTC output layer;
- add CTC-only training for ASR2 (I use this and the previous one to load a pretrained LM head into the CTC output layer and prealign encoder output with the pretraind LM embeddings);
- add `--pretrained_model` option to the ASR2 recipe to load the prealigned encoder parameters;
- allow dimensionality projection in the length adaptor to happen after the subsampling instead of before the subsampling (this needs less parameters and seems to work better).

To do:

- [x] add tests
- [x] add recipe

akreal commented:
@sw005320 this should be ready for a review (the CI failures are not related to my changes).

The recipe is based on LibriHeavy Small and Mistral-7B-Instruct-v0.2. The results actually show a regression in WER (15.4%&rarr;20.8% for test_other), but a different configuration might demonstrate LM benefits:
 - use the Mixtral-8x7B-v0.1 LM;
 - use the LibriHeavy Medium data;
 - fine-tune or jointly train on speech translation data (e.g. CoVoST2) or Dynamic-SUPERB.

Unfortunately, I don't have access to enough compute to try these configurations.

sw005320 commented:
Thanks a lot, @akreal!
@pengchengguo and @simpleoier, can you review this PR?

simpleoier commented on egs2/libriheavy_small/asr2/local/run_hf.sh:
```diff
@@ -0,0 +1,96 @@
+    mistralai/Mistral-7B-Instruct-v0.2 \
+    ${embed_tokens_path}
+
+./asr2.sh \
```
Can you add some comments why running asr2.sh twice?

akreal replied:
Done! Please let me know if I should clarify it more.

simpleoier commented on espnet2/mt/espnet_model.py:
```diff
@@ -63,6 +64,7 @@ def __init__(
         self.src_eos = src_vocab_size - 1 if src_vocab_size else None
         self.vocab_size = vocab_size
         self.src_vocab_size = src_vocab_size
+        self.src_ignore_id = src_ignore_id
```
I'm afraid that this makes the previous models incompatible.

akreal replied:
I guess it would, if `src_ignore_id` would register as a torch model parameter and it would be missing from `state_dict` in the previous models. However, it appears to be only a class attribute and `torch.load` does not pay attention to it.

I tested the baseline model from this recipe (`espnet/akreal_lh_small_asr2_e_branchformer_wavlm_large_21_km2k_bpe_rm6k_bpe_ts3k_sp`) on a master code with these two extra lines in `espnet2/mt/espnet_model.py`, and it worked:
```
src_ignore_id: int = -1,
...
self.src_ignore_id = src_ignore_id
```

Please let me know if you meant some other type of incompatibility.

simpleoier reviewed (commented):
Thanks for adding this new support! I left some comments. Please let me know if there is any misunderstanding.

akreal commented:
> Thanks for adding this new support! I left some comments. Please let me know if there is any misunderstanding.

Thank you for your review! I addressed your comments. Please let me know if there are further questions or suggestions.
